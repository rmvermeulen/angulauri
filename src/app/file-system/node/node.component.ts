import { Component, Input, OnInit } from '@angular/core'

export type IFileData = { path: string; content: string }
export type IFolderData = { path: string; children: INodeData[] }
export type INodeData = IFileData | IFolderData

@Component({
  selector: 'app-node',
  styles: [
    `
      .file {
      }
      .folder {
      }
      .even {
        background-color: #fee;
      }
      .odd {
        background-color: #eee;
      }
      .node {
        margin: 8px;
        padding: 8px;
        /*
        /* border-radius: 8px; */ */
        color: darkgray;
        border: solid black 1px;
      }
    `,
  ],
  template: `
    <div *ngIf="data" [ngClass]="classes">
      <h2 [class]="'file'" *ngIf="getFile() as file">
        file
        <code>{{ file.path }}</code>
        contains:
        <pre>{{ file.content }}</pre>
      </h2>
      <div [class]="'folder'" *ngIf="getFolder() as folder">
        folder <code>{{ folder.path }}</code> contains:
        <app-node
          *ngFor="let child of folder.children"
          [data]="child"
          [depth]="depth + 1"
        >
        </app-node>
      </div>
    </div>
  `,
})
export class NodeComponent implements OnInit {
  @Input() public depth = 0
  @Input() public data?: INodeData
  public get classes() {
    const isEven = !!(this.depth % 2)
    return {
      node: true,
      even: isEven,
      odd: !isEven,
    }
  }
  constructor() {}

  ngOnInit(): void {}

  public getFile(): IFileData | undefined {
    return this.data && 'content' in this.data ? this.data : undefined
  }
  public getFolder(): IFolderData | undefined {
    return this.data && 'children' in this.data ? this.data : undefined
  }
}
