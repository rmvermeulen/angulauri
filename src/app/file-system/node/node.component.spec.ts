import { ComponentFixture, TestBed } from '@angular/core/testing'
import { INodeData, NodeComponent } from './node.component'

describe.each([
  undefined,
  {
    path: '/',
    content: 'a file',
  },
  {
    path: '/',
    children: [
      {
        path: 'icon.png',
        content: 'a file',
      },
      {
        path: 'icon.jpg',
        content: 'a file',
      },
    ],
  },
])('NodeComponent data=%p', (data: INodeData | undefined) => {
  let component: NodeComponent
  let fixture: ComponentFixture<NodeComponent>

  beforeAll(async () => {
    await TestBed.configureTestingModule({
      declarations: [NodeComponent],
    }).compileComponents()
  })

  beforeAll(() => {
    fixture = TestBed.createComponent(NodeComponent)
    component = fixture.componentInstance
    component.data = data
    fixture.detectChanges()
  })

  it('should create', () => {
    expect(component).toBeTruthy()
    expect(fixture.nativeElement).toMatchSnapshot()
    expect(fixture.nativeElement.querySelector('div')).toMatchSnapshot()
  })

  it('matches expected output', () => {
    expect(fixture).toMatchSnapshot()
  })

  it.skip('should render a .node', () => {
    const node = fixture.nativeElement.querySelector('.node')
    expect(node).toBeTruthy()
  })

  it.skip('should render a .file', () => {
    const file = fixture.nativeElement.querySelector('.file')
    expect(file).toBeTruthy()
  })

  it.skip('should render a .folder', () => {
    const folder = fixture.nativeElement.querySelector('.folder')
    expect(folder).toBeTruthy()
  })
})
