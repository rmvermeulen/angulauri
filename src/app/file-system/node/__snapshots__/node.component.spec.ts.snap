// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeComponent data={"children": [Array], "path": "/"} matches expected output 1`] = `
<app-node
  data={[Function Object]}
  depth="0"
>
  <div
    class="node odd"
  >
    
    
    <div
      class="folder"
    >
       folder 
      <code>
        /
      </code>
       contains: 
      
      <app-node>
        
        <div
          class="node even"
        >
          
          <h2
            class="file"
          >
             file 
            <code>
              icon.png
            </code>
             contains: 
            <pre>
              a file
            </pre>
          </h2>
          
        </div>
      </app-node>
      <app-node>
        
        <div
          class="node even"
        >
          
          <h2
            class="file"
          >
             file 
            <code>
              icon.jpg
            </code>
             contains: 
            <pre>
              a file
            </pre>
          </h2>
          
        </div>
      </app-node>
    </div>
  </div>
</app-node>
`;

exports[`NodeComponent data={"children": [Array], "path": "/"} should create 1`] = `
<div
  id="root2"
>
  
  <div
    class="node odd"
  >
    
    
    <div
      class="folder"
    >
       folder 
      <code>
        /
      </code>
       contains: 
      
      <app-node>
        
        <div
          class="node even"
        >
          
          <h2
            class="file"
          >
             file 
            <code>
              icon.png
            </code>
             contains: 
            <pre>
              a file
            </pre>
          </h2>
          
        </div>
      </app-node>
      <app-node>
        
        <div
          class="node even"
        >
          
          <h2
            class="file"
          >
             file 
            <code>
              icon.jpg
            </code>
             contains: 
            <pre>
              a file
            </pre>
          </h2>
          
        </div>
      </app-node>
    </div>
  </div>
</div>
`;

exports[`NodeComponent data={"children": [Array], "path": "/"} should create 2`] = `
<div
  class="node odd"
>
  
  
  <div
    class="folder"
  >
     folder 
    <code>
      /
    </code>
     contains: 
    
    <app-node>
      
      <div
        class="node even"
      >
        
        <h2
          class="file"
        >
           file 
          <code>
            icon.png
          </code>
           contains: 
          <pre>
            a file
          </pre>
        </h2>
        
      </div>
    </app-node>
    <app-node>
      
      <div
        class="node even"
      >
        
        <h2
          class="file"
        >
           file 
          <code>
            icon.jpg
          </code>
           contains: 
          <pre>
            a file
          </pre>
        </h2>
        
      </div>
    </app-node>
  </div>
</div>
`;

exports[`NodeComponent data={"content": "a file", "path": "/"} matches expected output 1`] = `
<app-node
  data={[Function Object]}
  depth="0"
>
  <div
    class="node odd"
  >
    
    <h2
      class="file"
    >
       file 
      <code>
        /
      </code>
       contains: 
      <pre>
        a file
      </pre>
    </h2>
    
  </div>
</app-node>
`;

exports[`NodeComponent data={"content": "a file", "path": "/"} should create 1`] = `
<div
  id="root1"
>
  
  <div
    class="node odd"
  >
    
    <h2
      class="file"
    >
       file 
      <code>
        /
      </code>
       contains: 
      <pre>
        a file
      </pre>
    </h2>
    
  </div>
</div>
`;

exports[`NodeComponent data={"content": "a file", "path": "/"} should create 2`] = `
<div
  class="node odd"
>
  
  <h2
    class="file"
  >
     file 
    <code>
      /
    </code>
     contains: 
    <pre>
      a file
    </pre>
  </h2>
  
</div>
`;

exports[`NodeComponent data=undefined matches expected output 1`] = `
<app-node
  data="undefined"
  depth="0"
>
  
</app-node>
`;

exports[`NodeComponent data=undefined should create 1`] = `
<div
  id="root0"
>
  
</div>
`;

exports[`NodeComponent data=undefined should create 2`] = `null`;
